### PR TITLE
adds CONTRIBUTING document to the main Swift repository 

### PR DESCRIPTION
**Pull request by heckj**
Assisting the Swift OSS contributor experience team. Most of the content is migrated/replicated from https://www.swift.org/contributing/. I also pulled in some of the content from README.md to reference the huge detail for getting started with your first contribution and making an initial pull request.

I intentionally restructured the content to a sentence-per-line (as it's a bit easier to review, comment, and update in Pull Requests). Some latin-isms (.e.g.) were removed and replaced with plain English, and a few sentences converted from future-hopeful-tense to present-tense and active voice.

cc @ahoppen @parispittman

There's additional detail about toolchains, referenced from in here, in the README - that would likely be a good candidate for a separate article and on-par with the "getting started", talking to creating a toolchain from this and related repositories and using them from Xcode.

### CODE REVIEW

**Comment by parispittman** [comment]
@heckj , I think this looks great! Once we have this reviewed and merged in, we'll remove the text from the main guide and start the next part of the journey. :)

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+## Contributing Code
+
+If you are interested in:
+
+- Contributing fixes and features to the compiler: See our
+  [How to Submit Your First Pull Request guide](/docs/HowToGuides/FirstPullRequest.md).
+- Building the compiler as a one-off: See our [Getting Started guide][].
+- Building a toolchain as a one-off: Follow the [Getting Started guide][]
+  up until the "Building the project" section. After that, follow the
+  instructions in the [Swift Toolchains](/README.md#swift-toolchains).
```
Comment (ahoppen) [inline CONTRIBUTING.md:55]: I don’t think that a new contributor would want to build an entire toolchain. Doing so is its own can of worms and it’s rarely required. So I’d just remove the bullet. (You should be able to inline the Getting started guide link again as well.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+
+We also host answers to [Frequently Aasked Questions](/docs/HowToGuides/FAQ.md) that may be of interest.
```
Comment (ahoppen) [inline CONTRIBUTING.md:57]:
```suggestion
We also host answers to [Frequently Asked Questions](/docs/HowToGuides/FAQ.md) that may be of interest.
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+- Large or invasive changes usually have secondary changes that must be made before the large change (for example, API cleanup or addition).
+Commit these changes before the major change, independently of that work.
+
+- If possible, decompose the remaining interrelated work into unrelated sets of changes.
+Next, define the first increment and get consensus on the development goal of the change.
+
+- Make each change in the set either stand alone (for example, to fix a bug) or part of a planned series of changes that work toward the development goal.
+Explaining these relationships to the community can be helpful.
+
+If you are interested in making a large change and feel unsure about its overall effect, please make sure to first discuss the change and reach a consensus through the developer forums.
```
Comment (ahoppen) [inline CONTRIBUTING.md:92]:
```suggestion
If you are interested in making a large change and feel unsure about its overall effect, please make sure to first discuss the change and reach a consensus through the https://forums.swift.org.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+This tag helps email filters and searches for post-commit reviews.
+- When there is a body, separate it from the title by an empty line.
+- Make body concise, while including the complete reasoning.
+Unless required to understand the change, additional code examples or other details should be left to bug comments or the mailing list.
+- If the commit fixes an issue in the bug tracking system, include a link to the issue in the message.
+- For text formatting and spelling, follow the same rules as documentation and in-code comments—for example, the use of capitalization and periods.
+- If the commit is a bug fix on top of another recently committed change, or a revert or reapply of a patch, include the Git revision number of the prior related commit, for example “Revert abcdef because it caused bug#”.
+
+For minor violations of these guidelines, the community normally favors reminding the contributor of this policy over reverting.
+Minor corrections and omissions can be handled by sending a reply to the commits mailing list.
```
Comment (ahoppen) [inline CONTRIBUTING.md:117]: I have no idea what this sentence is supposed to refer to – we don’t have a commits mailing list and if we do there is no significant traffic on it. Let’s delete it.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+When contributors submit a change to a Swift subproject, after that change is approved, other developers with commit access may commit it for the author.
+When doing so, it is important to retain correct attribution of the contribution. Generally speaking, Git handles attribution automatically.
+
+We do not want the source code to be littered with random attributions like “this code written by J. Random Hacker”, which is noisy and distracting.
+Do not add contributor names to the source code or documentation.
+
+In addition, don’t commit changes authored by others unless they have submitted the change to the project or you have been authorized to submit on their behalf—for example, you work together and your company authorized you to contribute the changes.
+The author should first either submit the change through a pull request to the relevant project, email the development list, or add a bug tracker item.
+If someone sends you a change privately, encourage them to submit it to the appropriate list first.
```
Comment (ahoppen) [inline CONTRIBUTING.md:129]: I don’t know how ancient this paragraph it but most of it isn’t relevant with git and GitHub. I think the only relevant part is

```md
Don’t commit changes authored by others unless they have submitted the change to the project or you have been authorized to submit on their behalf — for example, you work together and your company authorized you to contribute the changes.
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+
+### Code Templates
+
+Code Templates
+
+As mentioned in the Community Overview, the license and copyright protections for Swift.org code are called out at the top of every source code file.
```
Comment (ahoppen) [inline CONTRIBUTING.md:135]: There is no *Community Overview* section in this document.

```suggestion
The license and copyright protections for Swift.org code are called out at the top of every source code file.
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+On the rare occasion you contribute a change that includes a new source file, ensure that the header is filled out appropriately.
```
Comment (ahoppen) [inline CONTRIBUTING.md:136]: I don’t think it’s that rare to contribute a new source file…

```suggestion
If you contribute a change that includes a new source file, ensure that the header is filled out appropriately.
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+The bottom section contains an optional description intended for generated documentation (these lines begin with `///` rather than `//`).
+If there is no description, this area can be skipped.
+
+### Code Review
+
+The Swift project relies heavily on code review to improve software quality:
+
+- All significant changes, by all developers, must be reviewed before they are committed to the repository.
+Smaller changes (or changes where the developer owns the component) can be reviewed after being committed.
+- Code reviews are conducted on GitHub (through comments on pull requests or commits) and are reflected on the relevant project’s commit mailing list.
```
Comment (ahoppen) [inline CONTRIBUTING.md:186]:
```suggestion
- Code reviews are conducted on GitHub.
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+- The developer responsible for a code change is also responsible for making all necessary review-related changes.
+
+Code review can be an iterative process, which continues until the change is ready to be committed.
```
Comment (ahoppen) [inline CONTRIBUTING.md:189]:
```suggestion
Code review can be an iterative process, which continues until the change is ready to be merged.
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+Please keep them short.
+
+### Quality
+
+People depend on Swift to create their production software.
+This means that a bug in Swift could cause bugs in thousands, even millions of developers’ products.
+Because of this, the Swift project maintains a high bar for quality.
+The minimum quality standards that any change must satisfy before being committed to the main development branch include:
+
+1. Code must compile without errors or warnings on at least one platform.
```
Comment (ahoppen) [inline CONTRIBUTING.md:221]:
```suggestion
1. Code must compile without errors on all supported platforms and should be free of warnings on at least one platform.
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+- Fix any major performance regressions.
+- Fix any performance or correctness regressions in the downstream Swift tools.
+- Fix any performance or correctness regressions that result in customer code that uses Swift.
+- Address any bugs that appear in the bug tracker as a result from your change.
+
+We prefer that these issues be handled before submission, but we understand that it isn’t possible to test all of this for every submission.
+Our continuous integration (CI) infrastructure normally finds these problems.
+We recommend watching the CI infrastructure throughout the next day to look for regressions.
+The CI infrastructure will directly email you if a group of commits that included yours caused a failure.
+You are expected to check those messages to see whether they are your fault and, if so, fix the breakage.
```
Comment (ahoppen) [inline CONTRIBUTING.md:239]: I don’t think this paragraph contributes to anything. CI testing is required for a PR to be merged and I don’t think it’s reasonable for anyone to *watch the CI infrastrucute* throughout the next day. Also, the CI infrastructure doesn’t send failure emails. Let’s just remove it.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+
+- You are granted commit-after-approval to all parts of Swift.
+To get approval, create a pull request. When the pull request is approved, you may merge it yourself.
+
+- You may commit an obvious change without first getting approval.
+The community expects you to use good judgment.
+Examples are reverting obviously broken patches, correcting code comments, and other minor changes.
+
+- You are allowed to commit changes without approval to the portions of Swift to which you have contributed or for which you have been assigned responsibility.
+Such commits must not break the build.
```
Comment (ahoppen) [inline CONTRIBUTING.md:279]: CI testing prevents you from breaking the build. Let’s remove this.

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+
+Swift follows a policy of making a change in the most upstream repository that is feasible.
+Contributions to Swift that involve Apple’s version of LLVM Project should go directly into the upstream LLVM repository unless they are specific to Swift.
+For example, an improvement to LLDB’s data formatters for a Swift type belongs in the Apple LLVM Project repository, whereas a bug fix to an LLVM optimizer—even if it’s only been observed when operating on Swift-generated LLVM IR—belongs in upstream LLVM.
+
+Commits to an upstream LLVM repository are automatically merged into the appropriate upstream branches in the corresponding Swift repository (`next`) in the [llvm-project](https://github.com/apple/llvm-project).
+
+### Swift and LLVM Developer Policies
+
+Contributions to Swift’s LLVM or Clang clones are governed by the [LLVM Developer Policy](http://llvm.org/docs/DeveloperPolicy.html) and should follow the appropriate [licensing](http://llvm.org/docs/DeveloperPolicy.html#copyright-license-and-patents) and [coding standards](http://llvm.org/docs/CodingStandards.html).
```
Comment (ahoppen) [inline CONTRIBUTING.md:313]:
```suggestion
Contributions to [llvm-project clone](https://github.com/swiftlang/llvm-project) are governed by the [LLVM Developer Policy](http://llvm.org/docs/DeveloperPolicy.html) and should follow the appropriate [licensing](http://llvm.org/docs/DeveloperPolicy.html#copyright-license-and-patents) and [coding standards](http://llvm.org/docs/CodingStandards.html).
```

**Review by ahoppen — COMMENTED** [review]
Thank you @heckj. Some sections are a little rusty and out-dated but it’s a great start. I added review comments for some of the ancient practices that no longer exist.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+If the issue is a crash, include a stack trace.
+Otherwise, describe the behavior you were expecting to see, along with the behavior you actually observed.
+
+- **A reproducible test case**.
+Double-check that your test case reproduces the issue.
+A relatively small sample (roughly within 50 lines of code) is best pasted directly into the description; a larger one may be uploaded as an attachment.
+Consider reducing the sample to the smallest amount of code possible; a smaller test case is easier to reason about and more appealing to сontributors.
+
+- **A description of the environment that reproduces the problem**.
+Include information about the Swift compiler’s version, the deployment target (if explicitly set) and your platform.
```
[the remark this answers is not in the document]
Comment (heckj) [reply CONTRIBUTING.md:33]: It would probably be better as a github issue template as well

[the remark this answers is not in the document]
Comment (ahoppen) [reply CONTRIBUTING.md:33]: That’s what I meant: We have GitHub issue forms that ask for all this information individually.

**Comment by ahoppen** [comment]
@heckj I think there are still a few open comments of mine. Can you let me know when you’re addressed them?

File: CONTRIBUTING.md (hunk 12 above)
Comment (heckj) [reply CONTRIBUTING.md]: Should be gone now with the latest commit

File: CONTRIBUTING.md (hunk 11 above)
Comment (heckj) [reply CONTRIBUTING.md]: Should be gone now with the latest commit

**Comment by heckj** [comment]
@ahoppen Ah, sorry - I added the last commit to remove the two additional sections you called out, but didn't resolve those conversations. Wasn't entirely sure of the etiquette there and if I should, or if that would be rude/inappropriate. In either case, the last commit update should have all your comments/suggestions addressed.

**Comment by ahoppen** [comment]
I think you haven’t addressed https://github.com/swiftlang/swift/pull/75654#discussion_r1706174061 yet, for example.

**Comment by heckj** [comment]
I removed the line "Such commits must not break the build." but left the rest in place. Did I misunderstand your suggestion - were you suggesting that the whole second bullet there?
> - You may commit an obvious change without first getting approval.

**Comment by ahoppen** [comment]
Sorry, I think I copied the wrong link. I updated my previous comments to that all that I think haven’t been addressed yet are marked as “Unresolved”.

**Comment by heckj** [comment]
I think that's sorted - thank you for all the suggestions in the diff format for tweaks and updates!

**Comment by ahoppen** [comment]
I think you missed a couple comments that were collapsed behind the “9 hidden conversations”, eg. https://github.com/swiftlang/swift/pull/75654#discussion_r1706167329

File: CONTRIBUTING.md (hunk 1 above)
Comment (heckj) [reply CONTRIBUTING.md:42]: removed bullet and inlined the link to the getting started docs

File: CONTRIBUTING.md (hunk 3 above)
Comment (heckj) [reply CONTRIBUTING.md]: make this a link, not just a raw URL

File: CONTRIBUTING.md (hunk 4 above)
Comment (heckj) [reply CONTRIBUTING.md:104]: dropping both sentences

File: CONTRIBUTING.md (hunk 5 above)
Comment (heckj) [reply CONTRIBUTING.md:116]: removed the non-github pieces and left in the "don't submit the work of others" bit with the caveat

**Comment by heckj** [comment]
Yep, definitely did. Sorry for the back and forth on this, there were quite a few I missed in there.

**Review by ahoppen — APPROVED** [review]
Thank you!

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by parispittman** [comment]
Thanks @heckj - and now we are on to the next step of removing this content from the main swift.org/contributing guide. onward! :)

**Pull request merged**